app/addons/documents/doc-editor/actions.js

```diff
@@ -156,75 +149,88 @@ function uploadAttachment (params) {
-          FauxtonAPI.addNotification({
-            msg: 'Document saved successfully.',
-            type: 'success',
-            clear: true
-          });
-        }.bind(this)
-      });
+  const query = '?rev=' + params.rev;
+  const db = params.doc.getDatabase().safeID();
+  const docId = params.doc.safeID();
+  const file = params.files[0];
+  const url = FauxtonAPI.urls('document', 'attachment', db, docId, encodeURIComponent(file.name), query);
 
-    },
-    error: function (resp) {
-      // cancelled uploads throw an ajax error but they don't contain a response. We don't want to publish an error
-      // event in those cases
-      if (_.isEmpty(resp.responseText)) {
-        return;
-      }
+  const onProgress = (evt) => {
+    if (evt.lengthComputable) {
+      const percentComplete = evt.loaded / evt.total * 100;
       FauxtonAPI.dispatch({
-        type: ActionTypes.FILE_UPLOAD_ERROR,
+        type: ActionTypes.SET_FILE_UPLOAD_PERCENTAGE,
         options: {
-          error: resp.responseJSON ? resp.responseJSON.reason : 'Error uploading file: (' + resp.statusText + ')'
+          percent: percentComplete
         }
       });
     }
-  });
+  };
+  const onSuccess = (doc) => {
+    // re-initialize the document editor. Only announce it's been updated when
+    initDocEditor({
+      doc: doc,
+      onLoaded: () => {
+        FauxtonAPI.dispatch({ type: ActionTypes.FILE_UPLOAD_SUCCESS });
+        FauxtonAPI.addNotification({
+          msg: 'Document saved successfully.',
+          type: 'success',
+          clear: true
+        });
+      }
+    });
+  };
+  const onError = (msg) => {
+    FauxtonAPI.dispatch({
+      type: ActionTypes.FILE_UPLOAD_ERROR,
+      options: {
+        error: msg
+      }
+    });
+  };
+  const httpRequest = new XMLHttpRequest();
+  currentUploadHttpRequest = httpRequest;
+  httpRequest.withCredentials = true;
+  if (httpRequest.upload) {
+    httpRequest.upload.onprogress = onProgress;
+  }
+  httpRequest.onloadend = () => {
+    currentUploadHttpRequest = undefined;
+  };
+  httpRequest.onerror = () => {
+    onError('Error uploading file');
+  };
+  httpRequest.onload = (e) => {
+    if (httpRequest.status >= 200 && httpRequest.status < 300) {
+      onSuccess(params.doc);
+    } else {
+      let errorMsg = 'Error uploading file. ';
+      if (e.responseText) {
+        try {
+          const json = JSON.parse(e.responseText);
+          if (json.error) {
+            errorMsg += 'Reason: ' + (json.reason || json.error);
+          }
+        } catch (err) {
+          //ignore parsing error
+        }
+      }
+      onError(errorMsg);
+    }
+  };
+  httpRequest.open('PUT', url);
+  httpRequest.setRequestHeader('X-Requested-With', 'XMLHttpRequest');
+  httpRequest.setRequestHeader('Content-Type', file.type || `application/octet-stream`);
+  httpRequest.setRequestHeader('Accept', 'application/json');
+  httpRequest.send(file);
```

Comment:
What is the advantage of using `XMLHttpRequest` over fetch to send the file?

Reply:
Reason was that we can't get the upload progress using fetch, so I had to use `XMLHttpRequest`.